Buy crypto by redirecting to onramp experience on pdapp instead of deposit popover

alaahd opened the pull request:
## Explanation
This pull request changes the implementation and the experience of buying crypto assets on MetaMask extension, which currently relies on hardcoded integration with a set of predefined providers (Coinbase Pay, Transak, MoonPay), to a brand-new implementation that improves the overall user experience by redirecting MetaMask users to the onramp curated experience on the MetaMask Portfolio Dapp (https://portfolio.metamask.io/buy).

With the new approach, we can achieve greater flexibility and better coverage by being able to streamline our operations dynamically and offer integration with new providers and crypto assets without the need for a new deployment, which would otherwise cause delays.

## Changes

- The previous hardcoded integration with predefined providers has been removed
- A new implementation that redirects the user to the MetaMask Portfolio Dapp for a better, curated buying experience has been added when clicking the buy button anywhere in the extension.
- The new implementation follows best practices for security and usability, ensuring a smooth and seamless user experience

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work? 

Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before

https://user-images.githubusercontent.com/2320312/220421951-4ec02c68-b057-4065-8763-6beff6f3f709.mov

<!-- How did the UI you changed look before your changes? Drag your file(s) below this line: -->


### After

https://user-images.githubusercontent.com/2320312/220422041-a9e9a15c-2dbd-473a-a3c9-232b28b679fd.mov

<!-- How does it look now? Drag your file(s) below this line: -->

## Manual Testing Steps

1. Navigate to the buying page
2. Initiate a buy action
3. Verify that you are redirected to the MetaMask Portfolio Dapp

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

alaahd committed:
feat: disable deposit popover and replace it with global redirect to onramp on pdapp

alaahd committed:
feat: remove legacy code for hardcoded onramp providers

bkirb commented:
@alaahd I verified that tapping Buy redirects the user to Portfolio. I tested while connected and not connected.

alaahd committed:
fix: remove unused visuals and components related to legacy code of deposit popover

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
fix: remove unused messages

alaahd committed:
feat: use a custom hook for all onramps related methods and variables

alaahd committed:
fix: modify the custom hook implementation to include test networks

alaahd committed:
fix: remove deprecated file buy-url

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
fix: remove references for deleted deposit logos

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
fix: network-controller failing unit test

alaahd committed:
fix: snapshot loading-swaps-quotes-stories-metadata.test.js.snap

alaahd committed:
fix: storybook tests

alaahd committed:
fix: remove unused constatns related to buyable onramp chains

alaahd committed:
fix: remove unused variables and fix eslint

alaahd committed:
Merge branch 'develop' into feat/buy-onramp-pdapp

alaahd committed:
Merge branch 'develop' into feat/buy-onramp-pdapp

alaahd committed:
Merge branch 'feat/buy-onramp-pdapp' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

jpuri commented on ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:
```diff
@@ -224,9 +217,12 @@ export default class ConfirmPageContainerContent extends Component {
+                          global.platform.openTab({
+                            url: `${portfolioUrl}/buy?metamaskEntry=ext_buy_button`,
+                          });
+                        }}
```
Hey @alaahd : we are working for increasing test coverage of confirmation pages, it will be nice to have above piece also tested. `global.platform.openTab` can probably be mocked for writing test.

jpuri commented on ui/components/app/confirm-page-container/confirm-page-container.component.js:
```diff
@@ -237,7 +241,17 @@ const ConfirmPageContainer = (props) => {
+                              text: 'Buy',
+                            },
+                          });
+                        }}
```
We try to keep this component for layout purpose only, any function invocation can be written [here](https://github.com/MetaMask/metamask-extension/blob/develop/ui/pages/confirm-transaction-base/confirm-transaction-base.component.js) and passed as prop to this component.

alaahd replied:
@jpuri The base component is a class component; knowing that I have created a custom hook to centralize the logic of redirecting to the Buy Crypto in portfolio dapp, how can we use this hook within a class component?

alaahd replied:
@jpuri this was the main reason I used the hook directly within `confirm-page-container.component.js` component, one example for a similar implementation is the usage of `useTransactionInsights`.

jpuri replied:
Make sense, can you plz add a comment about this.

We plan to make `confirm-transaction-base` also functional and once that is done we can extract this out.

alaahd replied:
I have addressed your feedback here https://github.com/MetaMask/metamask-extension/pull/17689/commits/fdb3df73299cedb243379336039b321186ce029a

jpuri commented:
Hey @alaahd : code changes look good. I added some feedback specific to confirmation components as we are working to improve their architecture.

Also, will be nice to have more test coverage in this PR.

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
adding unit test for useRamps custom hook

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

brad-decker reviewed (dismissed):
Wow. lots of code removal. well done.

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd committed:
feat: add comment on the proper usage of useRamps within confirm-page-container component

alaahd committed:
fix: add unit tests for buy button in token-overview page

alaahd committed:
fix: add unit test for open the buy crypto URL for a buyable chain ID in token page

alaahd committed:
feat: add unit test coverage for eth-overview page

alaahd committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into feat/buy-onramp-pdapp

alaahd commented:
> 



> Hey @alaahd : code changes look good. I added some feedback specific to confirmation components as we are working to improve their architecture.
> 
> Also, will be nice to have more test coverage in this PR.

@jpuri I have added more unit tests for the buy functionality. 

https://github.com/MetaMask/metamask-extension/blob/78893d85d747759006c59eac227ec821da4f437f/ui/components/app/wallet-overview/eth-overview.js

https://github.com/MetaMask/metamask-extension/blob/78893d85d747759006c59eac227ec821da4f437f/ui/components/app/wallet-overview/token-overview.test.js

alaahd committed:
fix: update locales

jpuri commented on ui/components/app/wallet-overview/eth-overview.js:
```diff
@@ -113,6 +113,7 @@ const EthOverview = ({ className }) => {
               <Icon name={ICON_NAMES.CARD} color={IconColor.primaryInverse} />
             }
             disabled={!isBuyableChain}
+            data-testid="eth-overview-buy"
```
Nit: we avoid using `data-testid`

brad-decker replied:
we don't anymore @jpuri -- it is *heavily* relied upon in e2e tests

jpuri commented on ui/pages/send/gas-display/gas-display.js:
```diff
@@ -339,7 +340,15 @@ export default function GasDisplay({ gasError }) {
+                              location: 'Gas Warning Insufficient Funds',
+                              text: 'Buy',
+                            },
+                          });
```
Nit: `trackEvent` inside `openBuyCryptoInPdapp` can save initialising `trackEvent`.

brad-decker replied:
Hmmm, guess you could pass the location/text into useRamps if that's the only thing that changes from one invocation to another. But if there are cases that you don't need to track an event i'd say take this into consideration for a future refinement.

alaahd replied:
@brad-decker I thought about doing this change, but then I said keeping the tracking code closer to the point of interaction (trigger) could be more efficient and traceable.

brad-decker reviewed (approved):
LGTM
